ortools/math_opt/solvers/xpress_solver.cc

```diff
@@ -0,0 +1,693 @@
+  const int num_coefficients = matrix.row_ids().size();
+  std::vector<int> row_index;
+  row_index.reserve(num_coefficients);
+  std::vector<int> col_index;
+  col_index.reserve(num_coefficients);
+  for (int k = 0; k < num_coefficients; ++k) {
+    row_index.push_back(
+        linear_constraints_map_.at(matrix.row_ids(k)).constraint_index);
+    col_index.push_back(variables_map_.at(matrix.column_ids(k)));
+  }
+  return xpress_->ChgCoeffs(row_index, col_index, matrix.coefficients());
+}
+
+absl::StatusOr<SolveResultProto> XpressSolver::Solve(
+    const SolveParametersProto& parameters,
+    const ModelSolveParametersProto& model_parameters,
+    MessageCallback message_cb,
+    const CallbackRegistrationProto& callback_registration, Callback cb,
+    const SolveInterrupter* interrupter) {
+  RETURN_IF_ERROR(ModelSolveParametersAreSupported(
+      model_parameters, kXpressSupportedStructures, "XPRESS"));
+  const absl::Time start = absl::Now();
+
+  RETURN_IF_ERROR(CheckRegisteredCallbackEvents(callback_registration,
+                                                /*supported_events=*/{}));
+
+  RETURN_IF_ERROR(CheckParameters(parameters));
+
+  // Check that bounds are not inverted just before solve
+  // XPRESS returns "infeasible" when bounds are inverted
+  {
+    ASSIGN_OR_RETURN(const InvertedBounds inv_bounds, ListInvertedBounds());
+    RETURN_IF_ERROR(inv_bounds.ToStatus());
+  }
+
+  // Set initial basis
+  if (model_parameters.has_initial_basis()) {
+    RETURN_IF_ERROR(SetXpressStartingBasis(model_parameters.initial_basis()));
+  }
+
+  RETURN_IF_ERROR(CallXpressSolve(parameters)) << "Error during XPRESS solve";
+
+  ASSIGN_OR_RETURN(SolveResultProto solve_result,
+                   ExtractSolveResultProto(start, model_parameters));
+
+  return solve_result;
+}
+
+inline std::string GetOptimizationFlags(
+    const SolveParametersProto& parameters) {
+  switch (parameters.lp_algorithm()) {
+    case LP_ALGORITHM_PRIMAL_SIMPLEX:
+      return "p";
+    case LP_ALGORITHM_DUAL_SIMPLEX:
+      return "d";
+    case LP_ALGORITHM_BARRIER:
+      return "b";
+    default:
+      // this makes XPRESS use default algorithm (XPRS_DEFAULTALG)
+      return "";
+  }
+}
+absl::Status XpressSolver::CallXpressSolve(
+    const SolveParametersProto& parameters) {
+  // Enable screen output right before solve
+  if (parameters.enable_output()) {
+    RETURN_IF_ERROR(xpress_->SetIntAttr(XPRS_OUTPUTLOG, 1))
+        << "Unable to enable XPRESS logs";
+  }
+  // Solve
+  if (is_mip_) {
+    ASSIGN_OR_RETURN(xpress_status_, xpress_->MipOptimizeAndGetStatus());
+  } else {
+    RETURN_IF_ERROR(SetLpIterLimits(parameters))
+        << "Could not set iteration limits.";
+    ASSIGN_OR_RETURN(xpress_status_, xpress_->LpOptimizeAndGetStatus(
+                                         GetOptimizationFlags(parameters)));
+  }
+  // Post-solve
+  if (!(is_mip_ ? (xpress_status_ == XPRS_MIP_OPTIMAL)
+                : (xpress_status_ == XPRS_LP_OPTIMAL))) {
```

Comment:
As I commented in the other file, the preferred way to solve is [XPRSoptimize()](https://www.fico.com/fico-xpress-optimization/docs/latest/solver/optimizer/HTML/XPRSoptimize.html) and then just check the solve and solution status returned by this function.